File: flight/OpenPilot/UAVObjects/uavobjecttemplate.c
/**
 ******************************************************************************
 *
 * @file       uavobjecttemplate.c
 * @author     The OpenPilot Team, http://www.openpilot.org Copyright (C) 2010.
 * @brief      Template file for all objects. This file will not compile, it is used
 * 	       by the parser as a template to generate all other objects. All $(x) fields
 * 	       will be replaced by the parser with the actual object information.
 * 	       Each object has a meta object associated with it. The meta object
 *             contains information such as the telemetry and logging properties.
 *
 * @see        The GNU Public License (GPL) Version 3
 *
 *****************************************************************************/
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#define $(METAOBJECT)

#include "openpilot.h"
#include "$(NAMELC).h"



// Private variables
UAVObjHandle handle;

/**
 * Initialize object.
 * \return 0 Success
 * \return -1 Failure
 */
int32_t $(NAME)Initialize()
{
	UAVObjMetadata metadata;

	// Register object with the object manager
	handle = UAVObjRegister($(NAMEUC)_OBJID, $(NAMEUC)_NAME, 0, $(NAMEUC)_SINGLEINST, $(NAMEUC)_NUMBYTES);
	if (handle == 0) return -1;

	// Initialize meta data
	metadata.ackRequired = $(ACK);
	metadata.gcsTelemetryUpdateMode = $(GCSTELEM_UPDATEMODE);
	metadata.gcsTelemetryUpdatePeriod = $(GCSTELEM_UPDATEPERIOD);
	metadata.telemetryUpdateMode = $(TELEM_UPDATEMODE);
	metadata.telemetryUpdatePeriod = $(TELEM_UPDATEPERIOD);
	metadata.loggingUpdateMode = $(LOGGING_UPDATEMODE);
	metadata.loggingUpdatePeriod = $(LOGGING_UPDATEPERIOD);
	UAVObjSetMetadata(handle, &metadata);

	// Done
	return 0;
}

UAVObjHandle $(NAME)GetHandle()
{
	return handle;
}

File: ground/openpilotgcs/src/libs/juavobjects/templates/uavobjecttemplate.java
/**
 ******************************************************************************
 *
 * @file       uavobjecttemplate.cpp
 * @author     The OpenPilot Team, http://www.openpilot.org Copyright (C) 2010.
 * @brief      Template for an uavobject in java
 *             $(GENERATEDWARNING)
 *             $(DESCRIPTION)
 *
 * @see        The GNU Public License (GPL) Version 3
 *
 *****************************************************************************/
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

package org.openpilot.uavtalk.uavobjects;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;

import org.openpilot.uavtalk.UAVObjectManager;
import org.openpilot.uavtalk.UAVObject;
import org.openpilot.uavtalk.UAVDataObject;
import org.openpilot.uavtalk.UAVObjectField;

/**
$(DESCRIPTION)

generated from $(XMLFILE)
 **/
public class $(NAME) extends UAVDataObject {

	public $(NAME)() {
		super(OBJID, ISSINGLEINST, ISSETTINGS, NAME);
		
		List<UAVObjectField> fields = new ArrayList<UAVObjectField>();
		
$(FIELDSINIT)

		// Compute the number of bytes for this object
		int numBytes = 0;
		ListIterator<UAVObjectField> li = fields.listIterator();
		while(li.hasNext()) {
			numBytes += li.next().getNumBytes();
		}
		NUMBYTES = numBytes;

		// Initialize object
		initializeFields(fields, ByteBuffer.allocate(NUMBYTES), NUMBYTES);
		// Set the default field values
		setDefaultFieldValues();
		// Set the object description
		setDescription(DESCRIPTION);
	}

	/**
	 * Create a Metadata object filled with default values for this object
	 * @return Metadata object with default values
	 */
	public Metadata getDefaultMetadata() {
		UAVObject.Metadata metadata = new UAVObject.Metadata();
    	metadata.flags =
		    UAVObject.Metadata.AccessModeNum(UAVObject.AccessMode.$(FLIGHTACCESS)) << UAVOBJ_ACCESS_SHIFT |
		    UAVObject.Metadata.AccessModeNum(UAVObject.AccessMode.$(GCSACCESS)) << UAVOBJ_GCS_ACCESS_SHIFT |
		    $(FLIGHTTELEM_ACKED) << UAVOBJ_TELEMETRY_ACKED_SHIFT |
		    $(GCSTELEM_ACKED) << UAVOBJ_GCS_TELEMETRY_ACKED_SHIFT |
		    UAVObject.Metadata.UpdateModeNum(UAVObject.UpdateMode.$(FLIGHTTELEM_UPDATEMODE)) << UAVOBJ_TELEMETRY_UPDATE_MODE_SHIFT |
		    UAVObject.Metadata.UpdateModeNum(UAVObject.UpdateMode.$(GCSTELEM_UPDATEMODE)) << UAVOBJ_GCS_TELEMETRY_UPDATE_MODE_SHIFT;
    	metadata.flightTelemetryUpdatePeriod = $(FLIGHTTELEM_UPDATEPERIOD);
    	metadata.gcsTelemetryUpdatePeriod = $(GCSTELEM_UPDATEPERIOD);
    	metadata.loggingUpdatePeriod = $(LOGGING_UPDATEPERIOD);
 
		return metadata;
	}

	/**
	 * Initialize object fields with the default values.
	 * If a default value is not specified the object fields
	 * will be initialized to zero.
	 */
	public void setDefaultFieldValues()
	{
$(INITFIELDS)
	}

	/**
	 * Create a clone of this object, a new instance ID must be specified.
	 * Do not use this function directly to create new instances, the
	 * UAVObjectManager should be used instead.
	 */
	public UAVDataObject clone(long instID) {
		// TODO: Need to get specific instance to clone
		try {
			$(NAME) obj = new $(NAME)();
			obj.initialize(instID, this.getMetaObject());
			return obj;
		} catch  (Exception e) {
			return null;
		}
	}

	/**
	 * Static function to retrieve an instance of the object.
	 */
	public $(NAME) GetInstance(UAVObjectManager objMngr, long instID)
	{
	    return ($(NAME))(objMngr.getObject($(NAME).OBJID, instID));
	}

	// Constants
	protected static final long OBJID = $(OBJIDHEX)l;
	protected static final String NAME = "$(NAME)";
	protected static String DESCRIPTION = "$(DESCRIPTION)";
	protected static final boolean ISSINGLEINST = $(ISSINGLEINST) > 0;
	protected static final boolean ISSETTINGS = $(ISSETTINGS) > 0;
	protected static int NUMBYTES = 0;


}
